src/main/java/nwoolcan/model/brewery/warehouse/article/ArticleManager.java

```diff
@@ -0,0 +1,137 @@
+package nwoolcan.model.brewery.warehouse.article;
+
+import nwoolcan.model.utils.UnitOfMeasure;
+import nwoolcan.utils.Result;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manager for {@link Article} objects. It is used by tests and by the
+ * {@link nwoolcan.model.brewery.warehouse.Warehouse} in order to create {@link Article}, check the
+ * id of the {@link Article}, avoid repetitions and set name of the {@link Article}.
+ */
+public final class ArticleManager {
+
+    @Nullable private static ArticleManager instance;
+    private static final String ARTICLE_NOT_REGISTERED = "The article was not registered. You can not change its name.";
+    private static final String ARTICLE_WITH_NEW_NAME_ALREADY_REGISTERED = "Changing the name to this article would produce an article which already exists.";
+    private static int fakeId = -1;
+    private int nextAvailableId;
+    private Map<Article, Integer> articleToId;
+    private Map<Integer, Article> idToArticle;
+
+    private ArticleManager() {
+        nextAvailableId = 1;
+        articleToId = new HashMap<>();
+        idToArticle = new HashMap<>();
+    }
+    /**
+     * Returns the only instance of the {@link ArticleManager} using a singleton pattern.
+     * @return the only instance of the {@link ArticleManager} using a singleton pattern.
+     */
+    public static synchronized ArticleManager getInstance() {
+        if (instance == null) {
+            instance = new ArticleManager();
+        }
+        return instance;
+    }
+    /**
+     * Checks the consistency of the article.
+     * @param article to be checked.
+     * @return a boolean denoting whether the id is correct or not.
+     */
+    public synchronized boolean checkId(final Article article) {
+        return articleToId.containsKey(article) && article.getId().equals(articleToId.get(article));
+    }
+    /**
+     * Creates a misc {@link Article} and insert it into the {@link ArticleManager}. If it already exists, it will be returned.
+     * @param name of the {@link Article}.
+     * @param unitOfMeasure the {@link UnitOfMeasure} of the {@link Article}.
+     * @return a new {@link Article} if it does not exist in the {@link ArticleManager}, otherwise the existing one.
+     */
+    @SuppressWarnings("NullAway")
+    public synchronized Article createMiscArticle(final String name,
+                                              final UnitOfMeasure unitOfMeasure) {
+        Article article = new ArticleImpl(fakeId, name, unitOfMeasure);
+        if (!articleToId.containsKey(article)) {
+            int newId = nextAvailableId++;
+            article = new ArticleImpl(newId, name, unitOfMeasure);
```

Comment:
All three methods `createArticle` seems to do the same thing, except calling different contructors.
Consider a private method doing the common parts for DRY.

Reply:
actually I can't find a way to do it by myself because everything is different except the increment of the newId. a code proposal would be very helpful

Reply:
The thing is that you are using the article and its hash code as a unique identifier.
Maybe keeping a set of hashcodes instead of two maps? You can then check if the hashcode is present or not to know if using a new id or not.

Reply:
Or a set of article maybe better

Reply:
The problem of a single set is that you can't get an element from a set. This leads to be impossible to call .getId() on it. I am trying to reduce maps from one two to one.

Reply:
I tried and initially it seemed it should have worked but testcases didn't pass. I need the map idToArticle because when returning the article it has to be the one in the articleManager, not the fake one.

Reply:
now it works with only one map :)